lib/src/util/dart_type_utilities.dart

```diff
@@ -219,7 +219,32 @@ class DartTypeUtilities {
     return nodes;
   }
 
+  /// Return whether [leftType] and [rightType] are _definitely_ unrelated.
+  ///
+  /// For the purposes of this function, here are some "relation" rules:
+  /// * `dynamic` and `Null` are considered related to any other type.
+  /// * Two equal types are considered related, e.g. classes `int` and `int`,
+  ///   classes `List<String>` and `List<String>`,
+  ///   classes `List<T>` and `List<T>`, and type variables `A` and `A`.
+  /// * Two types such that one is more specific than the other, such as classes
+  ///   `List<dynamic>` and `Iterable<dynamic>`, and type variables `A` and `B`
+  ///   where `A extends B`. The rules of type specificity are documented
+  ///   [InterfaceType.isMoreSpecificThan](https://pub.dartlang.org/documentation/analyzer/latest/dart_element_type/InterfaceType/isMoreSpecificThan.html).
+  /// * Two types, each representing a class:
+  ///   * are related if they represent the same class, modulo type arguments,
+  ///     and each of their pair-wise type arguments are related, e.g.
+  ///     `List<dynamic>` and `List<int>`, and `Future<T>` and `Future<S>` where
+  ///     `S extends T`.
+  ///   * are unrelated if [leftType]'s supertype is [Object].
+  ///   * are related if their supertypes are equal, e.g. `List<dynamic>` and
+  ///     `Set<dynamic>`.
+  /// * Two types, each representing a type variable, are related if their
+  ///   bounds are related.
+  /// * Otherwise, the types are related.
+  // TODO(srawlins): typedefs :D
```

Comment:
I don't understand what this TODO means. Could you elaborate?

Reply:
Oh sorry. I can add more to the comment, but basically I don't think this lint handles typedefs. There certainly aren't any tests :(